### PR TITLE
ref(searchbox): quick pass on performance

### PR DESCRIPTION
**Pull request by JonasBa**
We have been tracking this fn for a while through a [transaction in Sentry](https://sentry.io/organizations/sentry/discover/results/?field=p75%28transaction.duration%29&field=p99%28transaction.duration%29&field=p100%28transaction.duration%29&name=SearchBox.getTagList&project=-1&query=title%3ASearchBar.getTagList&statsPeriod=14d&yAxis=p75%28transaction.duration%29&yAxis=p99%28transaction.duration%29&yAxis=p100%28transaction.duration%29) and it seems that for some users the execution time could be well above the 16ms budget which would result in dropped frames. 

This is just a quick initial pass that avoids a few unnecessary fn calls and tries to reuse some values, there is probably more to do here, but we can re-evaluate this after.

CleanShot 2022-11-01 at 13 55 48@2x

### CODE REVIEW

File: static/app/components/events/searchBar.tsx [hunk 1]
```diff
@@ -33,63 +33,60 @@ const SEARCH_SPECIAL_CHARS_REGEXP = new RegExp(
   'g'
 );
 
-const getFunctionTags = (fields: Readonly<Field[]>) =>
-  Object.fromEntries(
```
Comment (JonasBa) [inline static/app/components/events/searchBar.tsx:37]: I'm not sure, but I would expect fromEntries to execute another On loop which we can avoid entirely by just doing filtering inside .reduce and return the acc. This is the majority of the changes

**Review by k-fish — APPROVED** [review]
I don't know how much impact the iteration changes will have on the perf (and I'm curious if a few loops could remove those extreme percentile outlier), it looks like while the correlation exists when looking at a heatmap even the upper end of input (n) is only 11.76ms on average. 

Screen Shot 2022-11-01 at 2 51 14 PM

That being said, we should try this anyway no matter what since I'm curious of the result 👍

**Comment by JonasBa** [comment]
>I don't know how much impact the iteration changes will have on the perf (and I'm curious if a few loops could remove those extreme percentile outlier), it looks like while the correlation exists when looking at a heatmap even the upper end of input (n) is only 11.76ms on average.

Yeah same here, I'm mostly curious as to what happens here. I'd expect an improvement, just not sure exactly how much.

File: static/app/components/events/searchBar.tsx [hunk 2]
```diff
@@ -33,63 +33,75 @@ const SEARCH_SPECIAL_CHARS_REGEXP = new RegExp(
...
-          !isEquation(item.field) &&
-          !isCustomMeasurement(item.field)
-      )
-      .map(item => [
-        item.field,
-        {key: item.field, name: item.field, kind: FieldKind.FUNCTION},
-      ])
-  );
+const STATIC_FIELD_TAGS_SET = new Set(Object.keys(FIELD_TAGS));
+const getFunctionTags = (fields: Readonly<Field[]> | undefined) => {
```
Comment (JonasBa) [inline static/app/components/events/searchBar.tsx:37]: @k-fish I think this should yield the most improvement (assumings fields has some large N here) as before it was doing Object.keys + includes for each item it was iterating over

Comment (k-fish) [reply static/app/components/events/searchBar.tsx:37]: oh so maybe O(n^2) or O(n*m)? That could do it 🤔. 10000 items should take trivial amounts of time to iterate over as long as they aren't huge and are monomorphic.

Comment (JonasBa) [reply static/app/components/events/searchBar.tsx:37]: Yeah, n*m here, but for the Object.entries + map I suspect that is On^2

Comment (JonasBa) [reply static/app/components/events/searchBar.tsx:37]: I just put up a quick benchmark - https://jsbench.me/msl9ynj2sc/1 - I get about 70% slower run on fromEntries

**Pull request merged**